### PR TITLE
Add wrap function

### PR DESCRIPTION
**Pull request by BSteephenson**
Adds a line wrapping function that limits line length to a certain width
In response to issue #278 

Let me know if there's anything more I should do (more tests, documentation, etc). I'm glad to help out.

### CODE REVIEW

**Comment by stoeffel** [comment]
I will have a look at it on the WE.Can you add the documentation to the `README.markdown`? Thanks for the PR.

File: README.markdown [hunk 1]
```diff
@@ -267,6 +267,15 @@ lines("Hello\nWorld");
 // => ["Hello", "World"]
 ```
 
+#### wrap(str, width, seperator, cut) => string
```
Comment (stoeffel) [inline README.markdown]: You should document the complete API here (defaults and some more examples).

File: test/strings.js [hunk 2]
```diff
@@ -899,5 +899,10 @@ $(document).ready(function() {
     strictEqual(_.toBoolean("foo FALSE bar", null, /FALSE/), false);
     strictEqual(_.toBoolean(" true  "), true);
   });
-
+  test('Strings: wrap', function(){
+    equal(_.wrap("My name is", 2, '.', false), "My .name .is", 'works with width 2 and cut = false');
+    equal(_.wrap("My name is", 2, '.', true), "My. n.am.e .is", 'works with width 2 and cut = true');
+    equal(_.wrap("My name is", 3, '.', false), "My .name .is", 'works with width 3 and cut = true');
+    equal(_.wrap("My name is", 3), "My \nname \nis", 'Default parameters work');
```
Comment (stoeffel) [inline test/strings.js]: You should add more tests.
* edgecases
* defaults

Checkout the other tests.

File: wrap.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// For example, wrap("I am me", 2) => "I \nam \nme"
+
+// str       : defualt ""    :  input line
+// width     : defualt 75    :  width of column of text
+// seperator : defualt '\n'  :  character that seperates lines
+// cut       : defualt false :  should it start a new line in the middle of a word?
+
+
+module.exports = function wrap(str, width, seperator, cut){
+	str = str || "";
```
Comment (stoeffel) [inline wrap.js]: You should use [`makeStr`](https://github.com/epeli/underscore.string/blob/master/helper/makeString.js) here.

File: wrap.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Wrap
+
+// wraps a string by a certain width
+// For example, wrap("I am me", 2) => "I \nam \nme"
+
+// str       : defualt ""    :  input line
+// width     : defualt 75    :  width of column of text
+// seperator : defualt '\n'  :  character that seperates lines
+// cut       : defualt false :  should it start a new line in the middle of a word?
```
Comment (stoeffel) [inline wrap.js]: You can move the jsdoc to the readme. We don't use jsdoc at the moment.

**Comment by stoeffel** [comment]
Can you rebase?

**Comment by BSteephenson** [comment]
Rebased with a mocha test
Let me know if there's anything else I should do =)

**Comment by stoeffel** [comment]
If you add this test:
`equal(wrap("My name is", -1), "My name is", "Just return original line if width <= 0")`
You have 100% coverage.

Otherwise it looks good to me.

**Comment by BSteephenson** [comment]
Done and squashed.

File: wrap.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+module.exports = function wrap(str, width, seperator, cut){
+	str = makeString(str);
+	width = width || 75;
+	seperator = seperator || '\n';
+	cut = cut || false;
+	if(width <= 0){
+		return str;
+	}
+	else if(!cut){
+		words = str.split(" ");
```
Comment (stoeffel) [inline wrap.js]: How about we use `underscore.string/words' here?

File: wrap.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		for(var i = 0; i < words.length - 1; i++){
+			//preserve the spaces for all words except the last one
```
Comment (stoeffel) [inline wrap.js]: I'm actually not sure if we should preserve the space. This behaviour is somehow different than I would expect it to be.

**Comment by BSteephenson** [comment]
Let's talk about the specifics of what this is supposed to do.

wrap('this is a sentence', 4, \n, false)

What should this return? Right now it would return 'this \nis a \nsentence' which I'm realizing now doesn't make much sense. Would 'this\nis a\nsentence' be more obvious?

**Comment by stoeffel** [comment]
> Would 'this\nis a\nsentence' be more obvious?

Yes, I think so.

PHP's `wordwrap` also drops spaces at the end. http://us3.php.net/manual/en/function.wordwrap.php
We could add an option for this, but it already has 4 arguments...

**Comment by epeli** [comment]
Options object maybe. Could be done as backwards compatible with type checking.

**Comment by stoeffel** [comment]
:+1: for an options object I could think of some other options we could add in the future f.e. dontBreakHTMLTags

**Comment by BSteephenson** [comment]
Now wrap('this is a sentence', {width: 4, seperator: '\n', cut: false}) returns 'this\nis a\nsentence'

**Comment by stoeffel** [comment]
Options object looks good. Could you add an option for dropping trailing spaces as discussed above? And updat the readme? Thanks for your work

**Comment by BSteephenson** [comment]
I added a trailingSpaces options which, if true, fills each line with trailing spaces to make them at least `width` long (I'm not sure if this is what you meant). I also updated the Readme.

**Comment by stoeffel** [comment]
Sorry, I actually meant this:

> What should this return? Right now it would return 'this \nis a \nsentence' which I'm realizing now doesn't make much sense. Would 'this\nis a\nsentence' be more obvious?

Maybe call the option `preserveSpace`. Sry for the confusion.

**Comment by stoeffel** [comment]
but the trailingSpaces option is nice too :+1:

**Comment by stoeffel** [comment]
@BSteephenson Would be great if you would add the `preserveSpaces` (or take a better name) option so we can merge this.

Thanks

**Comment by BSteephenson** [comment]
Working on this atm. I am having preserveSpaces have a higher precedence over trailingSpaces.

File: README.markdown [hunk 7]
```diff
@@ -269,6 +269,25 @@ lines("Hello\nWorld");
 // => ["Hello", "World"]
 ```
 
+#### wrap(str, options) => string
+
+Splits a line `str` (default '') into several lines of size `options.width` (default 75) using a `options.seperator` (default '\n'). If `options.trailingSpaces` is true, make each line at least `width` long using trailing spaces. If `options.cut` is true, create new lines in the middle of words. If `options.trailingSpaces` is true, preserve the space that should be there at the end of a line.
```
Comment (stoeffel) [inline README.markdown]: > `options.trailingSpaces` is true, preserve the space that should be there at the end of a line.

This should be `options.preserveSpaces`

File: README.markdown [hunk 8]
```diff
@@ -269,6 +269,25 @@ lines("Hello\nWorld");
...
+wrap("Hello World", { width:5 })
+// => "Hello\nWorld"
+
+wrap("Hello World", { width:6, seperator:'.', cut:false, trailingSpaces: true })
+// => "Hello .World "
+
+wrap("Hello World", { width:5, seperator:'.', cut:true, trailingSpaces: true })
+// => "Hello. Worl.d    "
+
+wrap("Hello World", { width:5, seperator:'.', cut:false, preserveSpaces: true })
```
Comment (stoeffel) [inline README.markdown]: I would drop `cut:false` here and on line 280 since the default is `false` anyway

Comment (stoeffel) [inline README.markdown]: should we state that `preserveSpaces` has no influence if `cut:true`?

**Comment by stoeffel** [comment]
Besides my comments in the readme, this looks good to me. Thx a lot

**Comment by BSteephenson** [comment]
Corrected all the oops's in Readme
Let me know if you want me to squash all the commits into one

**Comment by stoeffel** [comment]
Squashing would be nice. I want have time today to check it one more time. But I will review it again in the beginning of next week and then merge it.
Thanks for your work.

**Comment by stoeffel** [comment]
Did check it again. Could you squash? I will merge it.

**Comment by BSteephenson** [comment]
Done =)

**Comment by stoeffel** [comment]
nice thanks a lot.

**Comment by cahrens** [comment]
@stoeffel when "wrap" was added, it wasn't added to the "exports" function as a conflict with underscore. I know mixing in underscore.string is now discouraged, but unfortunately we have a lot of old code written with the mixin paradigm. I am having difficulty upgrading underscore.string from 2.x due to the wrap conflict.

**Comment by stoeffel** [comment]
Hey @cahrens. Good catch. Would be awesome if you could open a PR to include `wrap` in the exports.
I want to merge #481 and #482 first, before I publish the next version (if you have time and willing to review these two PRs would be cool).

Thanks

**Comment by cahrens** [comment]
Sure, I will looking into it this weekend. I would like to add a test that checks for conflicting functions between underscore and underscore.string. I haven't looked at your test framework yet, but hope it's possible to add a test-only dependency on underscore.

**Comment by cahrens** [comment]
@stoeffel I see that underscore is already a development dependency. Can you point me to documentation on how to run the tests locally? I poked around but couldn't find anything.

**Pull request merged**